File: ctci/ch1/is_unique/is_unique.py
# O(n)
def is_unique(string):
    freq_map = {}
    for i in range(len(string)):
        if string[i] not in freq_map:
            freq_map[string[i]] = 1
        else:
            return False
    return True

# O(n log n)
def is_unique_sort(string):
    string = ''.join(sorted(string))
    for i in range(1, len(string)):
        if string[i] == string[i-1]:
            return False
    return True

# O (n^2)
def is_unique_compare(string):
    for i in range(len(string)):
        for j in range(len(string)):
            if i != j and string[i] == string[j]:
                return False
    return True


print "Is unique: "
print is_unique("swag")
print is_unique("swags")
print is_unique("dogd")
print

print "Is unique sort: "
print is_unique_sort("swag")
print is_unique_sort("swags")
print is_unique_sort("dogd")
print

print "Is unique compare: "
print is_unique_compare("swag")
print is_unique_compare("swags")
print is_unique_compare("dogd")

File: ctci/ch1/is_unique/is_unique.go
package main

import "fmt"
import "sort"

func is_unique(x string) bool {
	freq_map := make(map[rune]int)
	for _, char := range x {
		freq_map[char] += 1
		if freq_map[char] > 1 {
			return false
		}
	}
	return true
}

func is_unique_compare(x string) bool {
	for i, elem1 := range x {
		for j, elem2 := range x {
			if i != j && elem1 == elem2 {
				return false
			}
		}
	}
	return true
}

func is_unique_sort(x string) bool {
	// Create an array of runes
	var r []rune
	for _, rune := range x {
		r = append(r, rune)
	}

	// fmt.Println(r)
	// Sort the array of runes
	sort.Slice(r, func(i, j int) bool {
		return r[i] < r[j]
	})

	// fmt.Println(r)

	// Convert the array of runes to a string
	sorted_string := string(r)
	// fmt.Println(sorted_string)

	for i := 0; i < len(sorted_string)-1; i++ {
		if sorted_string[i] == sorted_string[i+1] {
			return false
		}
	}

	return true
}

func main() {
	fmt.Println("Is unique: ")
	fmt.Println(is_unique("swag"))
	fmt.Println(is_unique("swags"))
	fmt.Println(is_unique("dogd"))
	fmt.Println()

	fmt.Println("Is unique sort: ")
	fmt.Println(is_unique_sort("swag"))
	fmt.Println(is_unique_sort("swags"))
	fmt.Println(is_unique_sort("dogd"))
	fmt.Println()

	fmt.Println("Is unique compare: ")
	fmt.Println(is_unique_compare("swag"))
	fmt.Println(is_unique_compare("swags"))
	fmt.Println(is_unique_compare("dogd"))
	fmt.Println()

}
